File: BOJ/Gold III/BOJ2143.py
#BOJ2143 두 배열의 합 20210411
import sys
input = sys.stdin.readline

def getSum(li,n):
    sums = [[] for _ in range(n)]
    for i in range(n):
        sums[i].append(li[i])
        for j in range(i+1,n):
            sums[i].append(sums[i][-1] + li[j])
    ret = []
    for v in sums: ret.extend(v)
    return ret

def main():
    t,n = [int(input()) for _ in range(2)]
    a = list(map(int,input().rstrip().split()))
    m = int(input())
    b = list(map(int,input().rstrip().split()))
    sumA = getSum(a,n)
    sumB = getSum(b,m)
    dicA = {}
    for v in sumA:
        if v not in dicA:
            dicA[v] = 1
        else:
            dicA[v] += 1
    cnt = 0
    for v in sumB:
        if t-v in dicA:
            cnt += dicA[t-v]
    print(cnt)
if __name__ == '__main__':
    main()

File: BOJ/Gold III/BOJ2143.java
//BOJ2143 두 배열의 합 20210706
import java.io.*;
import java.util.*;

import static java.lang.Math.*;

public class Main {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        int t = Integer.parseInt(br.readLine());
        int n = Integer.parseInt(br.readLine());
        int[] a = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int m = Integer.parseInt(br.readLine());
        int[] b = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        List<Integer> sumA = getSubSums(a, n);
        List<Integer> sumB = getSubSums(b, m);

        Map<Integer, Integer> sumACnt = new HashMap<>();
        for(int v: sumA){
            if(sumACnt.containsKey(v)){
                sumACnt.put(v, sumACnt.get(v) + 1);
            }
            else{
                sumACnt.put(v, 1);
            }
        }
        long ans = 0;
        for(int v: sumB){
            if(sumACnt.containsKey(t-v)){
                ans += sumACnt.get(t-v);
            }
        }
        System.out.println(ans);
    }

    public static List<Integer> getSubSums(int[] arr, int size){
        List<Integer> ret = new ArrayList<>();
        for(int i = 0; i < size; i++){
            int sum = 0;
            for(int j = i; j < size; j++){
                sum+= arr[j];
                ret.add(sum);
            }
        }
        return ret;
    }
}